src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java

```diff
@@ -93,46 +126,169 @@
-
-            final DetailAST classDef = findContainingClass(ast);
-            final DetailAST classMods = classDef.findFirstToken(TokenTypes.MODIFIERS);
-            // check if the containing class can be subclassed
-            final boolean classCanBeSubclassed = classDef.getType() != TokenTypes.ENUM_DEF
-                    && !classMods.branchContains(TokenTypes.FINAL);
-
-            if (nonEmptyImplementation && classCanBeSubclassed
-                    && hasDefaultOrExplicitNonPrivateCtor(classDef)) {
-
-                final String name = ast.findFirstToken(TokenTypes.IDENT).getText();
-                log(ast.getLineNo(), ast.getColumnNo(), MSG_KEY, name);
+        if (canBeOverridden(ast)
+                && !hasEmptyImplementation(ast)
+                && !hasJavadocComment(ast)
+                && !hasIgnoredAnnotation(ast, ignoredAnnotations)) {
+
+            final DetailAST containerDef = getNearestClassOrEnumDefinition(ast);
+            if (canBeSubclassed(containerDef)
+                    && hasDefaultOrExplicitNonPrivateCtor(containerDef)) {
+                final String methodName = ast.findFirstToken(TokenTypes.IDENT).getText();
+                log(ast.getLineNo(), ast.getColumnNo(), MSG_KEY, methodName);
+            }
+        }
+    }
+
+    /**
+     * Checks whether a method can be overridden in a subclass.
+     * Method can be overridden if it is not private, abstract, final or static.
+     * Note that the check has nothing to do for interfaces.
+     * @param methodDef method definition token.
+     * @return true if a method can be overridden in a subclass.
+     */
+    private boolean canBeOverridden(DetailAST methodDef) {
+        final DetailAST modifiers = methodDef.findFirstToken(TokenTypes.MODIFIERS);
+        return ScopeUtils.getSurroundingScope(methodDef).isIn(Scope.PROTECTED)
+            && !ScopeUtils.isInInterfaceOrAnnotationBlock(methodDef)
+            && !modifiers.branchContains(TokenTypes.LITERAL_PRIVATE)
+            && !modifiers.branchContains(TokenTypes.ABSTRACT)
+            && !modifiers.branchContains(TokenTypes.FINAL)
+            && !modifiers.branchContains(TokenTypes.LITERAL_STATIC);
+    }
+
+    /**
+     * Checks whether a method has an empty implementation.
+     * Method is OK if its implementation is empty.
+     * Note: native methods don't have implementation in java code, so
+     * implementation can be null even if method is not abstract.
+     * @param ast method definition token.
+     * @return true if a method has an empty implementation.
+     */
+    //-@cs[CyclomaticComplexity] We have to keep the complex logic
+    // as the method implements tree traverse algorithm.
+    private static boolean hasEmptyImplementation(DetailAST ast) {
+        boolean hasEmptyBody = true;
+
+        final DetailAST implementation = ast.findFirstToken(TokenTypes.SLIST);
+        final boolean isNativeMethod = implementation == null;
```

Comment:
native methods in java are marked by special modifier. native method != empty method (looks like , but not the same). Please use checking by modifier.
https://docs.oracle.com/javase/specs/jls/se8/html/jls-8.html#jls-8.4.3.4